File: codes/design/serialize_deserialize_btree.py
from collections import deque

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        ret = ""
        if not root:
            return ret

        q = deque()
        q.append(root)

        while q:
            t = q.popleft()
            if not t:
                ret += '$'
                ret += '#'
            else:
                ret += str(t.val)
                ret += '#'
                q.append(t.left)
                q.append(t.right)
        return ret

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if len(data) == 0:
            return None

        q = deque()
        pos = data.find('#')
        root = TreeNode(int(data[0:pos]))
        q.append(root)
        data = data[pos+1:]

        while q:
            t = q.popleft()
            pos = data.find('#')
            if data[0:pos] != '$':
                left_node = TreeNode(int(data[0:pos]))
                t.left = left_node
                q.append(left_node)
            data = data[pos+1:] # skip '#'

            pos = data.find('#')
            if data[0:pos] != '$':
                right_node = TreeNode(int(data[0:pos]))
                t.right = right_node
                q.append(right_node)
            data = data[pos+1:]

        return root

        
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: codes/design/serialize_deserialize_btree.cpp
#include "../base.h"

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class S_D_Btree_Recursive {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        ostringstream out;
        ser_helper(root, out);
        return out.str();
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        istringstream in(data);
        return des_helper(in);
    }
    
private:
    void ser_helper(TreeNode* root, ostringstream& out) {
        if (root) {
            out << root->val << " ";
            ser_helper(root->left, out);
            ser_helper(root->right, out);
        } else {
            out << "# ";
        }
    }
    
    TreeNode* des_helper(istringstream& in) {
        string val;
        if (!(in >> val))
            return nullptr;
        if (val == "#")
            return nullptr;
        TreeNode* node = new TreeNode(stoi(val));
        node->left = des_helper(in);
        node->right = des_helper(in);
        
        return node;
    }
};

class S_D_Btree {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        ostringstream out;
        if (!root)
            return out.str();
        
        queue<TreeNode*> q;
        q.push(root);
        
        while (!q.empty()) {
            TreeNode* t = q.front();
            q.pop();
            if (t) {
                out << t->val << ' ';
                q.push(t->left);
                q.push(t->right);
            } else {
                out << "# ";
            }
        }
        
        return out.str();
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int n = data.size();
        if (n == 0)
            return nullptr;
        istringstream in(data);
        
        string val;
        in >> val;      
        TreeNode* root = new TreeNode(stoi(val));
        queue<TreeNode*> q;
        q.push(root);
        
        while (!q.empty()) {
            TreeNode* t = q.front();
            q.pop();
            
            if (!(in >> val))
                break;
            if (val != "#") {
                TreeNode* left = new TreeNode(stoi(val));
                t->left = left;
                q.push(left);
            }
            
            if (!(in >> val))
                break;
            if (val != "#") {
                TreeNode* right = new TreeNode(stoi(val));
                t->right = right;
                q.push(right);
            }
        }
        
        return root;                                                                                                    
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));